app/Main.hs

```diff
@@ -99,13 +123,166 @@ makePscPackage force = do
+  T.writeTextFile "test/Main.purs" Templates.testMain
+  T.writeTextFile ".gitignore" Templates.gitignore
+  T.echo "Set up a local Spacchetti project."
+  T.echo "Try running `spacchetti install`"
+
+spacchettiJsonPath :: Text.Text
+spacchettiJsonPath = spacchettiDir <> "spacchetti.json"
+
+dhallToJsonCmd :: Text.Text
+dhallToJsonCmd = "cat ./spacchetti.dhall | dhall-to-json --pretty > " <> spacchettiJsonPath
+
+makeSpacchettiJson :: IO ()
+makeSpacchettiJson = do
+  T.mktree $ T.fromText spacchettiDir
+  code <- T.shell dhallToJsonCmd T.empty
+  case code of
+    T.ExitFailure n -> T.die ("making spacchetti.json failed: " <> T.repr n)
+    T.ExitSuccess -> pure ()
+
+readSpacchettiJson :: IO Types.SpacchettiConfig
+readSpacchettiJson = do
+  spacchettiConfigText <- T.readTextFile (T.fromText spacchettiJsonPath)
+  case eitherDecodeStrict $ Text.encodeUtf8 spacchettiConfigText of
+    Left e -> T.die $ "The generated spacchetti.json was in the wrong format: " <> Text.pack e
+    Right config -> return config
+
+install :: IO ()
+install = do
+  T.echo "generating spacchetti.json"
+  makeSpacchettiJson
+  spacchettiConfigText <- T.readTextFile (T.fromText spacchettiJsonPath)
+  config <- readSpacchettiJson
+  let deps = getAllDependencies config
+  echo' $ "installing " <> Text.pack (show $ List.length deps) <> " dependencies."
+  traverse getDep deps
+  T.echo "installation complete."
+
+ensureSpacchettiJson :: IO ()
+ensureSpacchettiJson = do
+  exists <- T.testfile $ T.fromText spacchettiJsonPath
+  T.unless exists makeSpacchettiJson
+
+getDir :: (Types.PackageName, Types.PackageDefinition) -> Text.Text
+getDir
+  ( Types.PackageName name
+  , Types.PackageDefinition
+      { Types.version=version
+      , Types.repo=repo
+      }
+  )
+  = spacchettiDir <> name <> "/" <> version
+
+getGlobs :: [(Types.PackageName, Types.PackageDefinition)] -> [Text.Text]
+getGlobs = map fn
+  where
+    fn pair = (getDir pair) <> "/src/**/*.purs"
+
+sources :: IO ()
+sources = do
+  ensureSpacchettiJson
+  config <- readSpacchettiJson
+  let
+    deps = getAllDependencies config
+    globs = getGlobs deps
+  _ <- traverse (echo') globs
+  pure ()
+
+getDep :: (Types.PackageName, Types.PackageDefinition) -> IO ()
+getDep
+  pair@
+    ( Types.PackageName name
+    , Types.PackageDefinition
+        { Types.version=version
+        , Types.repo=repo
+        }
+    ) = do
+  let dir = getDir pair
+  exists <- T.testdir $ T.fromText dir
+  if exists
+    then do
+      echo' $ name <> " already installed."
+    else do
+      echo' $ "installing " <> name
+      T.mktree . T.fromText $ dir
+      Dir.withCurrentDirectory (Text.unpack dir) $ do
+        let
+          cmd = Text.intercalate " && "
+            [ "git init"
+            , "git remote add origin " <> repo
+            , "git fetch origin " <> version
+            , "git -c advice.detachedHead=false checkout FETCH_HEAD"
+            ]
+        code <- T.shell cmd T.empty
```

Comment:
Not sure if we want to print out all `git` output?

Reply:
yeah, might as well hide it in the end